src/syft/lib/misc/union.py

```diff
@@ -0,0 +1,71 @@
+# stdlib
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Set
+from typing import Tuple
+
+
+class IndexableTrait(type):
+    """
+    IndexableTrait is used as a metaclass for our UnionGenerator to be able to
+    set __getitem__ on the class to enable the UnionGenerator[...] syntax.
+    """
+
+    def __getitem__(self, union_types: Tuple[str, ...]) -> str:
+        return UnionGenerator.from_qualnames(list(union_types))
+
+
+class UnionGenerator(type, metaclass=IndexableTrait):
+    """
+    UnionGenerator is a metaclass used to generate on the spot union types
+    from their qualnames.
+
+    Eg. UnionGenerator["syft.lib.python.Int", "syft.lib.python.Float"] will generate
+    in the current module the IntFloatUnion type that will have no functions added
+    to the type itself until the AST is finished to be generated. If new types of
+    unions have to be generated afterwards:
+        1. you might be using the ast in a wrong way.
+        2. if you know what you are doing, you have to regenarate the ast.
+    """
+
+    def __init__(self, *args: object):
+        super().__init__(self)
+
+    def __new__(cls, name: str, bases: tuple, dct: dict) -> Any:
+        dct["__qualname__"] = "syft.lib.misc.union." + name
+        dct["__name__"] = name
+        new_type = super().__new__(cls, name, bases, dct)
+        # insert our new union type in the global scope of the module
+        globals()[new_type.__name__] = new_type
+        return new_type
+
+    @staticmethod
+    def from_qualnames(union_types: List[str]) -> str:
+        # sorting to not generate the same permutation of the types
+        union_types = sorted(union_types)
+
+        name = (
+            "".join([union_type.split(".")[-1] for union_type in union_types]) + "Union"
+        )
+
+        # searching in the cache to not solve the union type multiple times
+        if name in union_cache:
+            return name
+
+        target_type = UnionGenerator(name, tuple(), {})
+        union_cache.add(name)
+
+        # adding the type and it's union types to a lazy dict that will be solved
```

Comment:
Q: Why is this a lazy dict? (it looks like a dict, right?)

Reply:
It's lazy in the sense that it delays the computation of the actual union types until you need to add them in the AST, not when requested. I a not sure I have better naming.